components/brave_vpn/brave_vpn_connection_manager_mac.mm

```diff
@@ -0,0 +1,266 @@
+
+// Referenced GuardianConnect implementation.
+// https://github.com/GuardianFirewall/GuardianConnect
+namespace brave_vpn {
+
+namespace {
+
+const NSString* kBraveVPNKey = @"BraveVPNKey";
+
+NSData* GetPasswordRefForAccount(const NSString* account_key) {
+  if (account_key == nil || [account_key length] == 0) {
+    LOG(ERROR) << "Error: account key is empty";
+    return nil;
+  }
+
+  NSString* bundle_id = [[NSBundle mainBundle] bundleIdentifier];
+  CFTypeRef copy_result = NULL;
+  NSDictionary* query = @{
+    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecAttrService : bundle_id,
+    (__bridge id)kSecAttrAccount : account_key,
+    (__bridge id)kSecMatchLimit : (__bridge id)kSecMatchLimitOne,
+    (__bridge id)kSecReturnPersistentRef : (__bridge id)kCFBooleanTrue,
+  };
+  OSStatus results =
+      SecItemCopyMatching((__bridge CFDictionaryRef)query, &copy_result);
+  if (results != errSecSuccess)
+    LOG(ERROR) << "Error: obtaining password ref(status:" << results << ")";
+  return (__bridge NSData*)copy_result;
+}
+
+OSStatus RemoveKeychainItemForAccount(const NSString* account_key) {
+  if (account_key == nil || [account_key length] == 0) {
+    LOG(ERROR) << "Error: account key is empty";
+    return errSecParam;
+  }
+
+  NSString* bundle_id = [[NSBundle mainBundle] bundleIdentifier];
+  NSDictionary* query = @{
+    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecAttrService : bundle_id,
+    (__bridge id)kSecAttrAccount : account_key,
+    (__bridge id)kSecReturnPersistentRef : (__bridge id)kCFBooleanTrue,
+  };
+  OSStatus result = SecItemDelete((__bridge CFDictionaryRef)query);
+  if (result != errSecSuccess && result != errSecItemNotFound)
+    LOG(ERROR) << "Error: deleting password entry(status:" << result << ")";
+  return result;
+}
+
+OSStatus StorePassword(const NSString* password, const NSString* account_key) {
+  if (password == nil || [password length] == 0) {
+    LOG(ERROR) << "Error: password is empty";
+    return errSecParam;
+  }
+
+  if (account_key == nil || [account_key length] == 0) {
+    LOG(ERROR) << "Error: account key is empty";
+    return errSecParam;
+  }
+
+  CFTypeRef result = NULL;
+  NSString* bundle_id = [[NSBundle mainBundle] bundleIdentifier];
+  NSData* password_data =
+      [[password dataUsingEncoding:NSUTF8StringEncoding] autorelease];
+  NSDictionary* sec_item = @{
+    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+    (__bridge id)kSecAttrService : bundle_id,
+    (__bridge id)kSecAttrSynchronizable : (__bridge id)kCFBooleanFalse,
+    (__bridge id)kSecAttrAccount : account_key,
+    (__bridge id)kSecValueData : password_data,
+  };
+  OSStatus status = SecItemAdd((__bridge CFDictionaryRef)sec_item, &result);
+  if (status != errSecSuccess) {
+    if (status == errSecDuplicateItem) {
+      VLOG(2) << "There is duplicated key in keychain. removing and re-adding.";
```

Comment:
Why would we remove and then re-add? Is there an advantage here im missing?

Reply:
You can use `SecItemUpdate` to fix this one if there is a duplicate.

Reply:
Changed to use `SecItemUpdate`.